### PR TITLE
Fix Box casting and sampling edge-cases

### LINKED ISSUE #768: [Bug Report] Box casting during init can cause invalid sampled values
### Describe the bug

During init, Box bounds are captured before casting to the destination dtype, but aren't checked again after the cast. 

When Box casts to a lower-precision float with a range smaller than the source dtype, for example `Box(0, 80000, dtype=np.float16)`, this can cause `high` and `low` to contain `np.inf` values without the corresponding `bounded_above` or `bounded_below` being False. This leads to an error when calling `sample()`.

When Box casts `np.inf` from a float to an int, for example `Box(255, np.inf, dtype=np.uint8)`, samples can under/overflow and be outside the Box bounds -- #328 seems to be related to this one.

### Code example

```shell
>>> from gymnasium.spaces import Box
>>> import numpy as np
>>> np.finfo(np.float16).max
65500.0
>>> b = Box(0, 80000, dtype=np.float16)
<snip>/python3.11/site-packages/numpy/core/numeric.py:330: RuntimeWarning: overflow encountered in cast
  multiarray.copyto(a, fill_value, casting='unsafe')
>>> b
Box(0.0, inf, (1,), float16)
>>> b.bounded_below
array([ True])
>>> b.bounded_above
array([ True])
>>> b.sample()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<snip>/python3.11/site-packages/gymnasium/spaces/box.py", line 228, in sample
    sample[bounded] = self.np_random.uniform(
                      ^^^^^^^^^^^^^^^^^^^^^^^
  File "_generator.pyx", line 1043, in numpy.random._generator.Generator.uniform
OverflowError: Range exceeds valid bounds
>>>
>>>
>>> ### New session
>>> 
>>> from gymnasium.spaces import Box
>>> import numpy as np
>>> b = Box(255, np.inf, dtype=np.uint8)
<snip>/python3.11/site-packages/numpy/core/numeric.py:330: RuntimeWarning: overflow encountered in cast
  multiarray.copyto(a, fill_value, casting='unsafe')
>>> b
Box(255, 255, (1,), uint8)
>>> b.bounded_below
array([ True])
>>> b.bounded_above
array([False])
>>> b.sample()
array([2], dtype=uint8)  # overflow!
```

### System info

- installed in a venv using pip
- `gymnasium.__version__ == '0.29.1'`
- macOS Ventura 13.5.2
- Python 3.11.3 (main, May 24 2023, 14:40:18) [Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin

### Additional context

_No response_

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/Farama-Foundation/Gymnasium/issues) in the repo

### PR DESCRIPTION
**Pull request by Jammf**
# Description

Changes Box to raise an error during init if either `low` or `high` are outside the limits of the Box's dtype. Values that are `np.inf`, `-np.inf`, or `np.nan` are ignored.

Also changes Box's sampling to prevent an edge-case. Having a `low` value near a dtype's _max_ int and an unbounded `high` value (or vice versa) would lead to an integer overflow (underflow) and invalid sampled values. Now the values are clipped before casting to ensure they're within the limits of the Box's dtype. For `np.int64`, it also clips again after casting, since the cast itself can cause the value to go outside the Box bounds for extreme values.

I also had to change a unit test for the DtypeObservation wrapper, since CartPole's observation space now cannot be cast to `np.int32` due to it using large (but finite) bounds outside the `np.int32` range.

Fixes #768 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have commented my code, particularly in hard-to-understand areas 
- [x] ~~I have made corresponding changes to the documentation~~
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by Jammf** [comment]
@pseudo-rnd-thoughts Looks like casting `np.inf` to a uint has undefined behavior, and macOS sets it to MAX_INT while the CI machine sets it to 0. There's special handling in `box._broadcast()` for signed ints, but nothing for uints. We can either define a behavior for uint (which would break backward compatibility for some platforms) or we can just disable those tests and leave things undefined?

**Comment by pseudo-rnd-thoughts** [comment]
@Jammf That seems pretty bad.

Before we make a change, can we be precise on the current functionally about this.
* Is low and high case to the `dtype` in `__init__`?
* What happens to `low` and `high` if outside the bounds of the `dtype` in `__init__`?
* What happens during `sample` if `low` and `high` are outside the bounds of the `dtype`?
* What is the value of `bounded_below` and `bounded_above` with `inf` and `val == dtype.max` (is the Box only bounded if inf is used or if the low or high is equal to the max value)? 

If you can think of any more relevant questions please add, would like to not mess up the Box space for old users where there isn't an issue already.
Then can be specify what of these answers we wish to change

**Comment by Kallinteris-Andreas** [comment]
@Jammf how are you casting `np.inf` to `integer`, that should result in an arithmetic error
```py
>>> import numpy as np
>>> int(np.inf)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OverflowError: cannot convert float infinity to integer
>>> np.inf.astype(np.int)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'float' object has no attribute 'astype'
```

**Comment by Jammf** [comment]
@Kallinteris-Andreas The relevant code is in `box._broadcast()`, and it looks like the logic for inf-to-int conversion was originally added in 37467417085788cd062347b5c0ea7a4e5ddf225d. Here's what's in main currently:

https://github.com/Farama-Foundation/Gymnasium/blob/8333df8666811d1d0f87f1ca71803cc58bcf09c6/gymnasium/spaces/box.py#L322-L329

(There's also similar logic for casting+broadcasting scalar `np.inf` and `-np.inf`)

It looks like `np.inf` and `-np.inf` are set to `MAX_INT - 2` and `MIN_INT + 2` for signed ints only. and uints were just left undefined.

With just numpy's casting, since `ndarray.astype` defaults to `unsafe` casting, it only emits a warning instead of raising an error:

```python
>>> import numpy as np
>>> a = np.array(np.inf)
>>> a.dtype
dtype('float64')

>>> a.astype(np.int16)
<stdin>:1: RuntimeWarning: invalid value encountered in cast
array(-1, dtype=int16)

>>> a.astype(np.int16, casting="safe")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Cannot cast scalar from dtype('float64') to dtype('int16') according to the rule 'safe'
```

**Comment by Jammf** [comment]
@pseudo-rnd-thoughts I spent some time looking into it the past few days, and it does seem pretty messy. I tried to condense the issues down to just the main points, but let me know anything doesn't make sense.

Core issues:
- out-of-range float `low`/`high` args act different from np.inf float `low`/`high` args
- out-of-range sint or uint `low`/`high` args can overflow, which causes errors during sampling
- -inf/inf sint `low`/`high` args can overflow when sampled
- inf uint `box.high` has different values depending on platform (and also can overflow when sampled)
- nans can be passed as `low`/`high`, which seems intentionally allowed by unit tests?

Other related discussions are at
- https://github.com/openai/gym/issues/2629
- https://github.com/openai/gym/issues/1929

### Current behavior

I made some charts to capture the current casting behavior in `main` (it doesn't cover any changes in this PR at all, since I'll probably end up needing to recommit new changes regardless). The first column is the input values (regardless of the dtype of the input arrays), the second column is the value of the `box.low` or `box.high` attributes, the third is the value of the `box.bounded_below` or `box.bounded_above` attributes, and the fourth is the sampling method and any special sampling issues or behavior.

| FLOATS       | attr. value      | bounded? | sampling     |
| ------------ | ---------- | -------- | ------------ |
| in-range     | (value)    | True     | Uniform      |
| out-of-range | inf        | True (!) | Uniform (!)  |
| inf/-inf     | inf/-inf   | False    | Exp/Normal   |
| nan          | nan        | False    |              |

| SIGNED INTS       | attr. value       | bounded? | sampling     |
| ----------------- | ---------- | -------- | ------------ |
| in-range          | (value)         | True     | Uniform        |
| out-of-range high | overflow value (!)   | True     | Uniform   |
| inf high          | MAX-2    | False    | Exp, can overflow (!)   |
| out-of-range low  | underflow value (!)  | True     | Uniform |
| -inf low          | MIN+2    | False    | Exp, can underflow (!) |
| nan               | 0 or MIN (depends on arch. and precision)   | False    |             |

| UNSIGNED INTS | attr. value       | bounded? | sampling     |
| ------------- | ------------ | -------- | ------------ |
| in-range          | (value)         | True     | Uniform         |
| out-of-range high  | overflow value (!)     | True     | Uniform |
| inf high      | MAX (arm64) or 0 (x86_64) | False    | Exp, can overflow (!) |
| out-of-range low  | underflow value (!)     | True     | Uniform |
| -inf low      | 0             | False    | Exp, can underflow (!) |
| nan           | 0             | False    |             |

Also, Exponential sampling is replaced with Normal if both high and low are unbounded, but I didn't include those cases. Normal sampling should lead to valid samples for floats and sints, and have a 50% chance of generating an invalid value for uints. Sampling from something as biased as a standard normal or exponential distribution doesn't seem ideal, but that's really a separate discussion.

As for potential fixes, I have a few ideas ordered by how much it'd break existing code. There's probably other solutions I missed, or mixes of different solutions would work. And of course, there's always the option to just do nothing.

### Proposal 1 

We could change uints/sints to be like how inf sints are currently handled, add clipping to sampling to prevent invalid samples, and raise errors for clearly nonsensical values (having nans, or `bounded_below=False` uints). This is the only solution that allows infinite-bounds integer boxes, by using magic values to represent inf and -inf. Those magic values could also be `MAX` and `MIN`  instead of `MAX-2` and `MIN+2`, as long as we're clipping properly during sampling.

| FLOATS       | attr. value      | bounded?    | sampling         |
| ------------ | ---------- | ----------- | ---------------- |
| in-range     | (value)    | True        | Uniform          |
| out-of-range | inf        | ***False*** | ***Exp/Normal*** |
| inf/-inf     | inf/-inf   | False       | Exp/Normal       |
| nan          | nan        | False       |                  |

| SIGNED INTS       | attr. value           | bounded? | sampling     |
| ----------------- | ------------- | -------- | ------------ |
| in-range          | (value)         | True     | Uniform        |
| out-of-range high | ***MAX-2*** | ***False*** | ***Exp, clipped to MAX-2*** |
| inf high          | MAX-2       | False       | Exp, ***clipped to MAX-2*** |
| out-of-range low  | ***MIN+2*** | ***False*** | ***Exp, clipped to MIN+2*** |
| -inf low          | MIN+2       | False       | Exp, ***clipped to MIN+2*** |
| nan               | ***raise Error*** | ***n/a*** | ***n/a***           |

| UNSIGNED INTS     | attr. value           | bounded? | sampling       |
| ----------------- | --------------- | -------- | -------------- |
| in-range          | (value)         | True     | Uniform        |
| out-of-range high | ***MAX-2*** | ***False*** | ***Exp, clipped to MAX-2*** |
| inf high          | ***MAX-2*** | False    | ***Exp, clipped to MAX-2*** |
| out-of-range low  | ***raise Error*** | ***n/a***  | ***n/a***        |
| -inf low          | ***raise Error*** | ***n/a*** | ***n/a***         |
| nan               | ***raise Error*** | ***n/a*** | ***n/a***         |

### Proposal 2
We could change infs and out-of-range ints to be like they were before d1f35fe5878b971441a5f0eea10bf6fb86fd1f0a, making it so they're always bounded. 

| FLOATS       | attr. value      | bounded? | sampling     |
| ------------ | ---------- | -------- | ------------ |
| out-of-range | inf        | ***False***| ***Exp/Normal*** |
| inf/-inf     | inf/-inf   | False    | Exp/Normal   |
| nan          | nan        | False    |              |

| SIGNED INTS  | attr. value        | bounded? | sampling     |
| ------------ | ------------ | -------- | ------------ |
| in-range          | (value)         | True     | Uniform        |
| out-of-range high | ***MAX*** | True  | ***Uniform*** |
| inf high     | ***MAX***    | ***True***    | ***Uniform*** |
| out-of-range low | ***MIN*** | True  | ***Uniform*** |
| -inf low     | ***MIN***    | ***True***    | ***Uniform*** |
| nan          | ***raise Error*** | ***n/a***    | ***n/a***           |

| UNSIGNED INTS     | attr. value           | bounded? | sampling       |
| ----------------- | --------------- | -------- | -------------- |
| in-range          | (value)         | True     | Uniform        |
| out-of-range high | ***MAX***     | True     | ***Uniform*** |
| inf high          | ***MAX***     | ***True*** | ***Uniform*** |
| out-of-range low  | ***raise Error*** | ***n/a***  | ***n/a***        |
| -inf low          | ***raise Error*** | ***n/a***  | ***n/a***      |
| nan               | ***raise Error*** | ***n/a***  | ***n/a***      |

### Proposal 3

We could raise an error if an out-of-range uint/sint is passed to `low`/`high`. This would force the caller to handle the casting.

| FLOATS       | attr. value      | bounded? | sampling     |
| ------------ | ---------- | -------- | ------------ |
| out-of-range | inf        | ***False***| ***Exp/Normal*** |
| inf/-inf     | inf/-inf   | False    | Exp/Normal   |
| nan          | nan        | False    |              |

| SIGNED INTS  | attr. value        | bounded? | sampling     |
| ------------ | ------------ | -------- | ------------ |
| in-range          | (value)         | True     | Uniform        |
| out-of-range high | ***raise Error*** | ***n/a***    | ***n/a*** |
| inf high     | ***raise Error*** | ***n/a***    | ***n/a*** |
| out-of-range low | ***raise Error*** | ***n/a***    | ***n/a*** |
| -inf low     | ***raise Error*** | ***n/a***    | ***n/a*** |
| nan          | ***raise Error*** | ***n/a***    | ***n/a*** |

| UNSIGNED INTS     | attr. value           | bounded? | sampling       |
| ----------------- | --------------- | -------- | -------------- |
| in-range          | (value)         | True     | Uniform        |
| out-of-range high | ***raise Error*** | ***n/a***    | ***n/a*** |
| inf high          | ***raise Error*** | ***n/a***    | ***n/a*** |
| out-of-range low  | ***raise Error*** | ***n/a***  | ***n/a***   |
| -inf low          | ***raise Error*** | ***n/a***  | ***n/a***   |
| nan               | ***raise Error*** | ***n/a***  | ***n/a***   |

### Proposal 4
We could make Box a continuous space only, plan to deprecate non-float Box, and redirect to the proper discrete spaces (MultiDiscrete and MultiBinary) instead. This would break existing pixel-observation envs that uses Box to encode image observations (and most of them probably do use Box), and would make a lot of tutorials be incorrect.

**Comment by pseudo-rnd-thoughts** [comment]
@Jammf Thank you for your highly detailed comment, this is as complex as I feared

I found the proposals a bit confusing so used your tables and some testing to propose what I think is the optimal solution that makes the most sense (and preserves backward compatibility)

## Proposed (optimal) solution

Looking at your proposals this is a bit different from all of yours with my primary change being to raise more errors in cases where sampling will most likely not make sense / work for users. What are your thoughts? 

| FLOATS          | attr value  | bounded?     | current sampling                 | proposed sampling  |
|-----------------|-------------|--------------|----------------------------------|--------------------|
| in-range        | value       | True         | uniform(low, high)               | uniform(low, high) | 
| -inf / in-range | inf / value | False / True | low + exponential                | low + exponential  |
| in-range / inf  | value / inf | True / False | high - exponential               | high - exponential | 
| -inf / inf      | -inf / inf  | False        | normal(0, 1)                     | normal(0, 1)       |
| out-of-range    | -inf / inf  | True         | uniform(low, high) with overflow | raise error!       |
| nan             | nan / nan   | False        | normal(0, 1)                     | raise error!       |

| SIGNED INTS     | attr value    | bounded?     | current sampling                 | proposed sampling  |
|-----------------|---------------|--------------|----------------------------------|--------------------|
| in-range        | value         | True         | uniform(low, high)               | uniform(low, high) | 
| -inf / in-range | MIN-2 / value | False / True | low + exponential                | low + exponential  |
| in-range / inf  | value / MAX-2 | True / False | high - exponential               | high - exponential | 
| -inf / inf      | MIN-2 / MAX-2 | False        | normal(0, 1)                     | normal(0, 1)       |
| out-of-range    | overflow!     | True         | uniform(low, high) with overflow | raise error!       |
| nan             | MIN-2 / MAX-2 | False        | normal(0, 1)                     | raise error!       |

| UNSIGNED INTS   | attr value | bounded?     | current sampling                 | proposed sampling  |
|-----------------|------------|--------------|----------------------------------|--------------------|
| in-range        | value      | True         | uniform(low, high)               | uniform(low, high) | 
| -inf / in-range | 0 / value  | False / True | low + exponential                | raise error!       |
| in-range / inf  | value / 0  | True / False | high - exponential               | raise error!       | 
| -inf / inf      | 0 / 0      | False        | normal(0, 1)                     | raise error!       |
| out-of-range    | overflow!  | True         | uniform(low, high) with overflow | raise error!       |
| nan             | nan / nan  | False        | normal(0, 1)                     | raise error!       |

The exponential sampling doesn't make much sense as the exponential distribution values (with scale=1.0) are so small compared to the range of values. However, due to backward compatibility, I don't think we can change this. 

We could add clipping at the end to ensure that samples are within the bounds but this is very unlikely, only possible with extreme values of one bound being inf and the other being very close to the bound value.

@Kallinteris-Andreas @jjshoots @RedTachyon do any of you have additional comments or thoughts?

**Comment by jjshoots** [comment]
I like @pseudo-rnd-thoughts's approach, makes it very verbose and doesn't silently allow the user to do something that the Box space was not supposed to do.
However, I'm not sure why we're raising errors only on sampling and not on init - for example, for signed ints, putting negative values or negative infinites should cause it to raise errors from the start.
Is this for legacy support?

**Comment by pseudo-rnd-thoughts** [comment]
@jjshoots apologies I didn't put in my comment that the error raising would happen all within the '__init__' as you noted. It was easiest to note that information in that column

**Comment by Kallinteris-Andreas** [comment]
My suggested changes of @pseudo-rnd-thoughts proposal

Foremost, the most important purposes of the spaces if to inform the learning algorithm what its observation/action spaces are, and secondly to sample an action space.

| FLOATS | attr value | bounded? | proposed sampling|
|-----------------|-------------|--------------|--------------------|
| in-range | value | True |uniform(low, high) |
| -inf / in-range | inf / value | False / True | low + exponential |
| in-range / inf | value / inf | True / False | high - exponential |
| -inf / inf | -inf / inf | False | normal(0, 1) |
| out-of-range | raise error! [1] | raise error! [1]  | raise error! |
| nan | raise error! [1] | raise error! [1]  | raise error! |



| SIGNED INTS | attr value | bounded? | proposed sampling |
|-----------------|---------------|--------------|--------------------|
| in-range | value | True | uniform(low, high) |
| -inf / in-range | MIN / value [3]| False / True | low + geometric [2] |
| in-range / inf | value / MAX [3]| True / False | high - geometric [2] |
| -inf / inf | MIN / MAX [3]| False | normal(0, 1) |
| out-of-range | raise error! [1] | raise error! [1]  | raise error! |
| nan | raise error! [1] | raise error! [1]  | raise error! |
    

    
| UNSIGNED INTS | attr value | bounded? | proposed sampling |
|-----------------|------------|--------------|--------------------|
| in-range | value | True |uniform(low, high) |
| -inf / in-range | raise error! [1] | raise error! [1] | raise error! | 
| in-range / inf | value / 0 | True / False | low + geometric [2, 4] | 
| -inf / inf | raise error! [1] | raise error! [1] |raise error! |
| out-of-range | raise error! [1] | raise error! [1] | raise error! |
| nan | raise error! [1] | raise error! [1] | raise error! |

[1] as @jjshots mentioned, this should be an error.
[2] exponential distribution does not exist for integers, [geometric](https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.geometric.html#numpy.random.Generator.geometric) should be used instead. (keep in mind, geometric distributions may still overflow).
[3] no reason to restrict the range to MIN-2/MAX-2 (even if still need it, it should be MIN+2).
[3b] or alternatively raise error for anything involving `inf` and integers, since it overcomplicates code, and only provides the benefit of different `sample()` method   
[4] I see no reason to not allow `uint`'s "in-range / inf" if `int`'s are also allowed.


>The exponential sampling doesn't make much sense as the exponential distribution values (with scale=1.0) are so small compared to the range of values. However, due to backward compatibility, I don't think we can change this.

To address this, we could add an optional argument (example name `sample_variable`) which would be normal distribution's `std`, exponential distribution's `scale` and geometric distribution's `p`.
(this could be a seperate PR)

**Comment by jjshoots** [comment]
I can't quite remember why I did min+2/max-2, but it was fairly important for some other part of the Box space. Perhaps we'll figure out why that's needed when we remove it and see if tests fail.

**Comment by pseudo-rnd-thoughts** [comment]
I suspect that the MAX-2 and MIN-2 was to account for `np.inf` and `np.nan` special values within the encoding I believe. 

My proposed sampling, building on @Kallinteris-Andreas solution is 

| FLOATS          | attr value   | bounded?     | proposed sampling           |
|-----------------|--------------|--------------|-----------------------------|
| in-range        | value        | True         | uniform(low, high)          |
| -inf / in-range | inf / value  | False / True | low + exponential(scaling)  |
| in-range / inf  | value / inf  | True / False | high - exponential(scaling) |
| -inf / inf      | -inf / inf   | False        | normal(0, scaling)          |
| out-of-range    | raise error! | raise error! | raise error!                |
| nan             | raise error! | raise error! | raise error!                |

| SIGNED INTS     | attr value   | bounded?     | proposed sampling       |
|-----------------|--------------|--------------|-------------------------|
| in-range        | value        | True         | uniform(low, high)      |
| -inf / in-range | MIN / value  | False / True | low + exponential(scale)  |
| in-range / inf  | value / MAX  | True / False | high - exponential(scale) |
| -inf / inf      | MIN / MAX    | False        | normal(0, scale)        |
| out-of-range    | raise error! | raise error! | raise error!            |
| nan             | raise error! | raise error! | raise error!            |
    
| UNSIGNED INTS   | attr value   | bounded?     | proposed sampling  |
|-----------------|--------------|--------------|--------------------|
| in-range        | value        | True         | uniform(low, high) |
| -inf / in-range | raise error! | raise error! | raise error!       | 
| in-range / inf  | raise error! | raise error! | raise error!       | 
| -inf / inf      | raise error! | raise error! | raise error!       |
| out-of-range    | raise error! | raise error! | raise error!       |
| nan             | raise error! | raise error! | raise error!       |

I would call the scaling variable, `sample_scaling: float = 1.0` within `__init__` to make it clear to users it's meaning and should be completed in another PR as @Kallinteris-Andreas said. 

For the geometric vs exponential, there seems to be minimal difference between the distribution (someone can correct but I believe that exponential is the continuous version of geometric distribution). So I'm not sure it makes sense to change

Looking at the tests that will need changing, could we add more testing for non-standard distributions, `np.bool_`, `np.uint8` and `np.float16`.

**Comment by Kallinteris-Andreas** [comment]
> removes unsigned integers in-range / inf

What is the reason for this choice

(other than that i agree)

**Comment by Jammf** [comment]
I found relatively few usages of bool Box in a [search of public GitHub repos](https://github.com/search?q=%2F%28%3F-i%29Box%5C%28.*bool%2F+language%3APython+AND+%28%22from+gymnasium%22+OR+%22import+gymnasium%22%29&type=code), so it'd probably only have minor impact to remove it or redirect to MultiBinary. (Though, the link only has single-line usages, I wasn't able to figure out how to get a multi-line regex to work nicely)

**Comment by jjshoots** [comment]
Crap thought I was looking at a different PR. My bad.

**Comment by pseudo-rnd-thoughts** [comment]
I will try to finish this PR by the end of the week. As the technical details are complex, I've decided to start with the tests to outline our expectations of what the space should be. This doesn't cover the sampling currently but that hasn't changed.

I found it easier to rewrite most of the tests to order them into different groupings: shape, dtype, low/high and infinite space

I will work on implementing the changes this evening hopefully

**Comment by pseudo-rnd-thoughts** [comment]
Over a month later, I've finished this PR by spending 5 hours on this and rewriting most of the Box init function. 
The result is a horribly complex function but adds the functionality discussed above: disallow nan, out of range values, unsigned int inf and -inf values. 

A large amount of the function is processing the `low` and `high` value where to check for out of range has to be completed before converted to the final dtype. 
However, this required specialised code for scalars and arrays, repeated for low and high (see `_cast_low` and `_cast_high`)

File: tests/spaces/test_box.py [hunk 1]
```diff
@@ -9,20 +9,54 @@
...
         ),
+        (
+            object,
+            ValueError,
+            "Invalid Box dtype (object), must be an integer, floating, or bool dtype",
+        ),
+        (
+            str,
+            ValueError,
+            "Invalid Box dtype (<U0), must be an integer, floating, or bool dtype",
```
[the remark this answers is not in the document]
Comment (Kallinteris-Andreas) [reply tests/spaces/test_box.py:41]: i am confused, but ok

File: gymnasium/spaces/box.py [hunk 2]
```diff
@@ -80,72 +80,214 @@ def __init__(
...
 
-        assert isinstance(low, np.ndarray)
-        assert (
-            low.shape == shape
-        ), f"low.shape doesn't match provided shape, low.shape: {low.shape}, shape: {shape}"
-        assert isinstance(high, np.ndarray)
-        assert (
-            high.shape == shape
-        ), f"high.shape doesn't match provided shape, high.shape: {high.shape}, shape: {shape}"
+        # most of the code between these two functions are copied with minor changes for < and >
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/spaces/box.py:147]: The next two function calls

File: gymnasium/spaces/box.py [hunk 3]
```diff
@@ -80,72 +80,214 @@ def __init__(
...
+            )
 
-        self.low_repr = _short_repr(self.low)
-        self.high_repr = _short_repr(self.high)
+        self.low_repr = array_short_repr(self.low)
+        self.high_repr = array_short_repr(self.high)
 
         super().__init__(self.shape, self.dtype, seed)
 
+    def _cast_low(self, low, dtype_min):
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/spaces/box.py:172]: Added

**Review by Kallinteris-Andreas — APPROVED** [review]

**Pull request merged**